pam/go-loader/module.c

```diff
@@ -0,0 +1,142 @@
+/* A simple PAM wrapper for GO based pam modules
+ *
+ * Copyright (C) 2023 Marco Trevisan
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General
+ * Public License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Marco Trevisan <marco.trevisan@canonical.com>
+ */
+
+#include <dlfcn.h>
+#include <limits.h>
+#include <security/pam_modules.h>
+#include <security/pam_ext.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+/* When a Go shared library is loaded from C, go starts various goroutine
+ * (as init() at first) and if the loading code is then performing a fork
+ * we end up having an undefined behavior and very likely, deadlocks.
+ *
+ * To avoid this we need to ensure that the module is called after the loading
+ * application has forked, and this for sure has happened when the application
+ * calls the PAM functions that a module exposes.
+ *
+ * As per this, here we re-implement the PAM modules functions and for each
+ * action we load the module if it has not been loaded already, otherwise we
+ * dload it and we redirect each call to the loaded library.
+ */
+
+typedef int (*PamHandler)(pam_handle_t *,
+                          int          flags,
+                          int          argc,
+                          const char **argv);
+
+static void
+on_go_module_removed (pam_handle_t *pamh,
+                      void         *go_module,
+                      int           error_status)
+{
+  void (*go_pam_cleanup) (void);
+  *(void **) (&go_pam_cleanup) = dlsym (go_module, "go_pam_cleanup_module");
```

Comment:
Do you expect `go_pam_cleanup_module` and `go_pam_init_module` to be in the finale pam Go module and executing them? What would they do?

My reasoning is: if those are just subs for now (as we even don’t fail if they don’t exists, just not executing them), let’s not add them until we really them.

Reply:
Ok, I added them as they could be useful. I had not many ideas for them. *Maybe* one option was to run `runtime.GC()` during cleanup, but this is not fully needed.

Ideally having a way to stop all the go threads would be cool, but I don't think there's one from what I searched.

Reply:
There is an internal way of doing this, (the testing framework is using it to mark the exit code for instance), but it’s not exposed outside of internal syscall per architecture.

The `runtime.GC()` is a good idea though.